### PR TITLE
function to fund other delegators

### LINKED ISSUE #117: Enable parties to delegate stake for others
As a user, I want to be able to delegate stake on behalf of others. I would like to achieve this via a function that looks similar to the below:

`public entry fun add_delegators(funder: &signer, delegator_addresses: vector<address>, principle_stakes: vector<u64>)`

`delegator_addresses` should specify the addresses for which the amount specified at the corresponding index of `principle_stakes` should be delegated. The sum of `principle_stakes` should be deducted from my account.

### PR DESCRIPTION
**Pull request by sjoshisupra**
Fixes #117

### CODE REVIEW

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move [hunk 1]
```diff
@@ -865,7 +866,33 @@ module supra_framework::pbo_delegation_pool {
             enable_partial_governance_voting(pool_address);
         }
     }
+    
+    public entry fun fund_delegators_with_locked_stake(funder: &signer, pool_address: address, delegators: vector<address>, stakes: vector<u64>) acquires DelegationPool, BeneficiaryForOperator, GovernanceRecords, NextCommissionPercentage {
 
+
+        let  pool = borrow_global_mut<DelegationPool>(pool_address);
+        
+        vector::zip_reverse(delegators,stakes,|delegator,stake| {
```
Comment (isaacdoidge) [inline aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Why are you reversing? Shouldn't `zip` be sufficient?

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move [hunk 2]
```diff
@@ -865,7 +866,33 @@ module supra_framework::pbo_delegation_pool {
...
+            table::add(&mut pool.principle_stake,delegator,stake);
+        }
+        });
+
+        fund_delegators_with_stake(funder,pool_address,delegators,stakes);
+    }
+
+    public entry fun fund_delegators_with_stake(funder: &signer, pool_address: address, delegators: vector<address>, stakes : vector<u64>) acquires DelegationPool, BeneficiaryForOperator, GovernanceRecords, NextCommissionPercentage {
+        //length equality check is performed by `zip_reverse`
+        vector::zip_reverse(delegators,stakes, |delegator,stake| {
```
Comment (isaacdoidge) [inline aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Again, why `reverse`?

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move [hunk 3]
```diff
@@ -395,6 +395,7 @@ module supra_framework::pbo_delegation_pool {
     }
 
     struct AddStakeEvent has drop, store {
+        stake_funder: address,
```
Comment (isaacdoidge) [inline aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Better not to add this here so that we can preserve backwards compatibility with testnet.

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move [hunk 4]
```diff
@@ -1298,14 +1322,22 @@ module supra_framework::pbo_delegation_pool {
         // in order to appreciate all shares on the active pool atomically
         buy_in_active_shares(pool, NULL_SHAREHOLDER, add_stake_fee);
 
+        supra_account::transfer(funder, pool_address, amount);
```
Comment (isaacdoidge) [inline aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Why did you change the order of this operation?

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move (hunk 3 above)
Comment (isaacdoidge) [reply aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: I think we can leave this information out of the event for now. As you said, coin transfer events will be emitted too, so we can use those to identify the funder if need be.

**Review by isaacdoidge — COMMENTED** [review]
Thanks for the quick work! I have a few questions.

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move (hunk 1 above)
Comment (sjoshisupra) [reply aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: @isaacdoidge , I am not reversing anything. `zip_reverse` would access both vectors from the end, using `pop` which would be cheaper than forward pass using `borrow`. During `pop` it would consume/destroy the vector, which is fine here.

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move (hunk 2 above)
Comment (sjoshisupra) [reply aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Same answer

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move (hunk 4 above)
Comment (sjoshisupra) [reply aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Yeah, this is fixed in new commit. Changing the order fails test because `pool` does not have sufficient fund to support `stake::add_stake`. Originally I thought this `transfer` would be in another function, but it turns out this is not needed. So order is fixed in the new commits.

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move (hunk 3 above)
Comment (sjoshisupra) [reply aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: fixed in added commits

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move (hunk 1 above)
Comment (isaacdoidge) [reply aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Thanks for explaining

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move [hunk 5]
```diff
@@ -856,16 +921,67 @@ module supra_framework::pbo_delegation_pool {
...
+        stakes: vector<u64>
+    ) acquires DelegationPool, BeneficiaryForOperator, GovernanceRecords, NextCommissionPercentage {
+        vector::zip_reverse(
+            delegators,
+            stakes,
+            |delegator, stake| {
+
+                //Compute the actual stake that would be added, `principle_stake` has to be
+                //populated in the table accordingly
+                let added_stake = stake - get_add_stake_fee(pool_address, stake);
```
Comment (isaacdoidge) [inline aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Have you tested what happens if `stake == 0`?

Comment (sjoshisupra) [reply aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Ignoring adding to table in the new commit. If the `fund_delegator_stake` has `amount==0`, then it doesn't do anything anyway. Essentially, this ignores zero stake addition.

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move (hunk 1 above)
Comment (ghost) [reply aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Nice solution!

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move [hunk 6]
```diff
@@ -4996,6 +5152,223 @@ module supra_framework::pbo_delegation_pool {
         assert!(coin::balance<SupraCoin>(new_delegator_address) == (100 * ONE_SUPRA) - 1, 0);
     }
 
+    #[test(supra_framework = @supra_framework, validator = @0x123, delegator = @0x010, funder=@0x999)]
+    #[expected_failure(abort_code = 10, location = Self)]
+    /// say unlocking schedule is 3 month cliff, monthly unlocking of 10% and principle stake is 100 coins then
+    /// between 3 and 4 months, check that it's can't unlock there principal stacke
```
Comment (ghost) [inline aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move:8047]:
```suggestion
    /// between 3 and 4 months, check that it's can't unlock there principal stake
```

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move [hunk 7]
```diff
@@ -4944,6 +4979,127 @@ module supra_framework::pbo_delegation_pool {
...
+
+        unlock(&new_delegator_address_signer, pool_address,(30 * ONE_SUPRA));
+        unlock(&new_delegator_address_signer2, pool_address,(70 * ONE_SUPRA));
+
+
+        timestamp::fast_forward_seconds(LOCKUP_CYCLE_SECONDS);
+        end_aptos_epoch();
+        withdraw(&new_delegator_address_signer, pool_address,(30 * ONE_SUPRA));
+        let new_delegator_balance = coin::balance<SupraCoin>(new_delegator_address);
+        assert!(new_delegator_balance == (30 * ONE_SUPRA) - 1 , new_delegator_balance);
```
Comment (ghost) [inline aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: Is this `-1` off because there are some numerical computation round error? If so, maybe we should add some comment

Comment (sjoshisupra) [reply aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: @axiongsupra , yes, it is off by one due to rounding. We have tested earlier that it doesn't go anywhere, it is indeed still available in the pool.

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move [hunk 8]
```diff
@@ -1298,6 +1327,7 @@ module supra_framework::pbo_delegation_pool {
         // in order to appreciate all shares on the active pool atomically
         buy_in_active_shares(pool, NULL_SHAREHOLDER, add_stake_fee);
 
+        let funder_address = signer::address_of(funder);
```
Comment (ghost) [inline aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: The funder_address is never used. Is there something missing?

File: aptos-move/framework/supra-framework/sources/stake.move [hunk 9]
```diff
@@ -2795,11 +2795,7 @@ module supra_framework::stake {
             let vci: &ValidatorConsensusInfo = obj;
             validator_consensus_info::get_addr(vci)
         });
-        let vci_pks = vector::map_ref(&vci_vec_0, |obj|{
```
Comment (ghost) [inline aptos-move/framework/supra-framework/sources/stake.move:2798]: Btw, what are those change for?

Comment (nizam-supraoracles) [reply aptos-move/framework/supra-framework/sources/stake.move:2798]: @axiongsupra, removed the `vci_pks` variable, as it appears to be unused.

**Review by nizam-supraoracles — APPROVED** [review]

File: aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move [hunk 10]
```diff
@@ -7869,12 +7870,184 @@ module supra_framework::pbo_delegation_pool {
...
+    }
+
+    #[
+    test(
+        supra_framework = @supra_framework,
+        validator = @0x123,
+        delegator = @0x010,
+    )
+    ]
+    #[expected_failure(abort_code = 65561)]
```
Comment (nizam-supraoracles) [inline aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move]: when you write `expected_failure` test, you need to define `location` with abort_code, just to remove compiler warning.

**Review by isaacdoidge — APPROVED** [review]
